bridge/engine/SsrcOutboundContext.cpp

```diff
@@ -1,33 +1,437 @@
+
+    if (!(needToRewriteLevelExtension || needToRewriteAbsSendTimeExtension))
+    {
+        return;
+    }
+
+    const auto headerExtensions = rtpHeader.getExtensionHeader();
+    if (!headerExtensions)
+    {
+        return;
+    }
+
+    for (auto& rtpHeaderExtension : headerExtensions->extensions())
+    {
+        if (needToRewriteLevelExtension &&
+            rtpHeaderExtension.getId() == senderInboundContext.rtpMap.audioLevelExtId.get())
+        {
+            if (rtpMap.audioLevelExtId.isSet())
+            {
+                rtpHeaderExtension.setId(rtpMap.audioLevelExtId.get());
+            }
+            else
+            {
+                rtpHeaderExtension.fillWithPadding();
+            }
+        }
+        else if (needToRewriteAbsSendTimeExtension &&
+            rtpHeaderExtension.getId() == senderInboundContext.rtpMap.absSendTimeExtId.get())
+        {
+            if (rtpMap.absSendTimeExtId.isSet())
+            {
+                rtpHeaderExtension.setId(rtpMap.absSendTimeExtId.get());
+            }
+            else
+            {
+                rtpHeaderExtension.fillWithPadding();
+            }
+        }
+    }
+}
+
+void SsrcOutboundContext::doRtpHeaderExtensionRewriteForVideo(rtp::RtpHeader& rtpHeader,
+    const bridge::SsrcInboundContext& senderInboundContext)
+{
+
+    const auto headerExtensions = rtpHeader.getExtensionHeader();
+    if (!headerExtensions)
+    {
+        return;
+    }
+
+    const bool senderHasAbsSendTimeEx = senderInboundContext.rtpMap.absSendTimeExtId.isSet();
+    const bool receiverHasAbsSendTimeEx = rtpMap.absSendTimeExtId.isSet();
+    const bool absSendTimeExNeedToBeRewritten = senderHasAbsSendTimeEx && receiverHasAbsSendTimeEx &&
+        senderInboundContext.rtpMap.absSendTimeExtId.get() != rtpMap.absSendTimeExtId.get();
+
+    if (absSendTimeExNeedToBeRewritten)
+    {
+        for (auto& rtpHeaderExtension : headerExtensions->extensions())
+        {
+            if (rtpHeaderExtension.getId() == senderInboundContext.rtpMap.absSendTimeExtId.get())
+            {
+                rtpHeaderExtension.setId(rtpMap.absSendTimeExtId.get());
+                return;
+            }
+        }
+    }
+}
+
+void SsrcOutboundContext::doRtpHeaderRewriteForAudio(rtp::RtpHeader& header,
+    const bridge::SsrcInboundContext& senderInboundContext,
+    const uint32_t newSequenceNumber,
+    const bool isTelephoneEvent)
 {
+    assert(!(isTelephoneEvent && telephoneEventRtpMap.isEmpty()));
+
+    header.sequenceNumber = newSequenceNumber & 0xFFFFu;
+    header.ssrc = this->ssrc;
+    header.timestamp = _rewrite.offset.timestamp + header.timestamp;
+    header.payloadType = isTelephoneEvent ? telephoneEventRtpMap.payloadType : rtpMap.payloadType;
+    doRtpHeaderExtensionRewriteForAudio(header, senderInboundContext);
+}
 
-bool SsrcOutboundContext::shouldSend(uint32_t ssrc, uint32_t sequenceNumber) const
```

Comment:
Deciding if the packet should be sent is now part of the rewrite code logic itself. All rewrite code was centralized in `SssrcOuboundContext` and the logic about whether  the rewrite is possible or not should not be scattered over multiple files as it makes the logic error prone